### PR TITLE
Integration/stockup consume interfaces

### LINKED ISSUE #176: Change implementation of Stockup command to use stockup model interface

### PR DESCRIPTION
**Pull request by rebstan97**
Closes #176

### CODE REVIEW

File: src/main/java/seedu/address/model/ingredient/UniqueIngredientList.java [hunk 1]
```diff
@@ -96,48 +96,65 @@ public Ingredient find(IngredientName ingredientName) throws IngredientNotFoundE
...
+        for (HashMap.Entry<IngredientName, Integer> ingredientPair : requiredIngredients.entrySet()) {
             IngredientName name = ingredientPair.getKey();
             Integer unitsToAdd = ingredientPair.getValue();
 
-            Ingredient ingredient = find(name);
+            Ingredient ingredient;
+            try {
+                ingredient = find(name);
+            } catch (IngredientNotFoundException e) {
+                consume(stockedUpIngredients);
```
Comment (AZhiKai) [inline src/main/java/seedu/address/model/ingredient/UniqueIngredientList.java]: I don't get this line. Is it consuming an ingredient that's not found?

File: src/main/java/seedu/address/model/ingredient/UniqueIngredientList.java [hunk 2]
```diff
@@ -96,48 +96,65 @@ public Ingredient find(IngredientName ingredientName) throws IngredientNotFoundE
...
+        for (HashMap.Entry<IngredientName, Integer> ingredientPair : requiredIngredients.entrySet()) {
             IngredientName name = ingredientPair.getKey();
             Integer unitsToConsume = ingredientPair.getValue();
 
-            Ingredient ingredient = find(name);
+            Ingredient ingredient;
+            try {
+                ingredient = find(name);
+            } catch (IngredientNotFoundException e) {
+                stockUp(consumedIngredients);
```
Comment (AZhiKai) [inline src/main/java/seedu/address/model/ingredient/UniqueIngredientList.java]: Same here...

File: src/main/java/seedu/address/model/ingredient/UniqueIngredientList.java (hunk 1 above)
Comment (rebstan97) [reply src/main/java/seedu/address/model/ingredient/UniqueIngredientList.java]: We want it such that when we stockup/consume a list of ingredients, as long as one ingredient fails, the whole operation should be abandoned. So, let's say the first ingredient is valid and stocked up, but the second ingredient does not exist, we throw an exception and undo the stock up operation for the first ingredient (equivalent to consuming first ingredient by the same units).

Comment (AZhiKai) [reply src/main/java/seedu/address/model/ingredient/UniqueIngredientList.java]: Ah I see, that makes sense. So it's an atomic operation -- all or nothing occurs. Okay, great. 👍

File: src/main/java/seedu/address/model/ingredient/UniqueIngredientList.java [hunk 3]
```diff
@@ -96,48 +96,65 @@ public Ingredient find(IngredientName ingredientName) throws IngredientNotFoundE
...
-     * Stocks up the ingredients which are keys in HashMap {@code recipe} by increasing the number of units of each
-     * ingredient.
+     * Stocks up the ingredients which are keys in HashMap {@code requiredIngredients} by increasing the number
+     * of units of each ingredient.
      * @throws  IngredientNotFoundException if the ingredient does not exist in the list.
      */
-    public void stockUp(HashMap<IngredientName, Integer> recipe) throws IngredientNotFoundException {
-        requireNonNull(recipe);
-        for (HashMap.Entry<IngredientName, Integer> ingredientPair : recipe.entrySet()) {
+    public void stockUp(HashMap<IngredientName, Integer> requiredIngredients) throws IngredientNotFoundException {
```
Comment (yican95) [inline src/main/java/seedu/address/model/ingredient/UniqueIngredientList.java]: The project use Interface as method parameter
`public Person(Name name, Phone phone, Email email, Address address, Remark remark, Set<Tag> tags) {`
I think maybe it is better to take Map as the parameter.
```suggestion
    public void stockUp(Map<IngredientName, Integer> requiredIngredients) throws IngredientNotFoundException {
```

Comment (AZhiKai) [reply src/main/java/seedu/address/model/ingredient/UniqueIngredientList.java]: I agree w Yi Can on this. Using interface here ensures minimal code changes in the event the implementation changes. Also, you can easily switch between the implementation of Map, should you wish to.

Comment (rebstan97) [reply src/main/java/seedu/address/model/ingredient/UniqueIngredientList.java]: Sounds logical, okay will make the change 👍

**Pull request merged**